The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12393079738/job/34593611885) [(plain)](https://github.com/rust-lang/rust/commit/398ffc035c66bfc1c9d6d87099523d4d7d947816/checks/34593611885/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    |
68  | fn walk_unsafe(
    |    ----------- this function can't have a `self` parameter
...
102 |             match self.body.exprs[*expr] {
    |                   ^^^^ `self` value is a keyword only available in methods with a `self` parameter
error[E0424]: expected value, found module `self`
   --> crates/hir-ty/src/diagnostics/unsafe_check.rs:107:21
    |
68  | fn walk_unsafe(
68  | fn walk_unsafe(
    |    ----------- this function can't have a `self` parameter
...
107 |                     self.body
    |                     ^^^^ `self` value is a keyword only available in methods with a `self` parameter
error[E0424]: expected value, found module `self`
   --> crates/hir-ty/src/diagnostics/unsafe_check.rs:108:70
    |
68  | fn walk_unsafe(
68  | fn walk_unsafe(
    |    ----------- this function can't have a `self` parameter
...
108 |                         .walk_child_exprs_without_pats(expr, |child| self.walk_expr(child));
    |                                                                      ^^^^ `self` value is a keyword only available in methods with a `self` parameter
   Compiling xshell-macros v0.2.7
   Compiling rust-analyzer v0.0.0 (/checkout/src/tools/rust-analyzer/crates/rust-analyzer)
    Checking xshell v0.2.7
    Checking process-wrap v8.0.2
```

</details>

--- a/src/tools/rust-analyzer/crates/hir-ty/src/diagnostics/unsafe_check.rs
+++ b/src/tools/rust-analyzer/crates/hir-ty/src/diagnostics/unsafe_check.rs
@@ -99,10 +99,18 @@ fn walk_unsafe(
         }
         Expr::Path(path) => mark_unsafe_path(path, current.into()),
         Expr::Ref { expr, rawness: Rawness::RawPtr, mutability: _ } => {
-            if let Expr::Path(_) = body.exprs[*expr] {
-                // Do not report unsafe for `addr_of[_mut]!(EXTERN_OR_MUT_STATIC)`,
-                // see https://github.com/rust-lang/rust/pull/125834.
-                return;
+            match body.exprs[*expr] {
+                Expr::Path(_) => return,
+                // https://github.com/rust-lang/rust/pull/129248
+                // Taking a raw ref to a deref place expr is always safe.
+                Expr::UnaryOp { expr, op: UnaryOp::Deref } => {
+                    body.walk_child_exprs(expr, |child| {
+                        walk_unsafe(db, infer, body, resolver, def, child, inside_unsafe_block, unsafe_expr_cb);
+                    });
+
+                    return;
+                }
+                _ => (),
             }
         }
         Expr::MethodCall { .. } => {

--- a/src/tools/rust-analyzer/crates/ide-diagnostics/src/handlers/missing_unsafe.rs
+++ b/src/tools/rust-analyzer/crates/ide-diagnostics/src/handlers/missing_unsafe.rs
@@ -631,4 +631,19 @@ fn main() {
 "#,
         );
     }
+
+    #[test]
+    fn raw_ref_reborrow_is_safe() {
+        check_diagnostics(
+            r#"
+fn main() {
+    let ptr: *mut i32;
+    let _addr = &raw const *ptr;
+    let local = 1;
+    let ptr = &local as *const i32;
+    let _addr = &raw const *ptr;
+}
+"#,
+        )
+    }
 }

